conjure/servicewriter.go

```diff
@@ -1018,25 +1049,20 @@ func returnTypesForEndpoint(endpointDefinition spec.EndpointDefinition, info typ
 	var returnTypes []expression.Type
 	imports := make(StringSet)
 	if endpointDefinition.Returns != nil {
-		var goType string
-		returnBinary, err := isReturnTypeSpecificType(endpointDefinition.Returns, visitors.IsBinary)
+		typer, err := visitors.NewConjureTypeProviderTyper(*endpointDefinition.Returns, info)
 		if err != nil {
 			return nil, nil, err
 		}
+		returnBinary, err := isReturnTypeSpecificType(endpointDefinition.Returns, visitors.IsBinary)
 		if returnBinary {
 			// special case: "binary" type resolves to []byte in structs, but indicates a streaming response when
 			// specified as the return type of a service, so use "io.ReadCloser".
-			goType = types.IOReadCloserType.GoType(info)
+			typer = types.MapBinaryTypeToReadCloserType(typer)
 			imports.AddAll(NewStringSet(types.IOReadCloserType.ImportPaths()...))
-		} else {
-			typer, err := visitors.NewConjureTypeProviderTyper(*endpointDefinition.Returns, info)
-			if err != nil {
-				return nil, nil, err
-			}
-			goType = typer.GoType(info)
-			if err != nil {
-				return nil, nil, errors.Wrapf(err, "failed to process return type %q", goType)
-			}
+		}
+		goType := typer.GoType(info)
+		if err != nil {
```

Comment:
If this is the `err` check that pairs with the `isReturnSpecificType` call, it should be moved up; otherwise, should be removed.